pkg/controller/config/config.go

```diff
@@ -0,0 +1,236 @@
+	if !ok {
+		antreaAgentConfig[types.DefaultMTUOption] = types.DefaultMTU
+	}
+
+	// Set Antrea image.
+	if operConfig.Spec.AntreaImage == "" {
+		operConfig.Spec.AntreaImage = types.DefaultAntreaImage
+	}
+	if operConfig.Spec.AntreaImage != types.DefaultAntreaImage {
+		log.Info("WARNING: expected Antrea Image is: %s, but got: %s", types.DefaultAntreaImage, operConfig.Spec.AntreaImage)
+	}
+
+	updatedAntreaAgentConfig, err := yaml.Marshal(antreaAgentConfig)
+	if err != nil {
+		return fmt.Errorf("failed to fill default serviceCIDR in antrea-agent config: %v", antreaAgentConfig)
+	}
+	operConfig.Spec.AntreaAgentConfig = string(updatedAntreaAgentConfig)
+	return nil
+}
+
+func ValidateConfig(clusterConfig *configv1.Network, operConfig *operatorv1.AntreaInstall) error {
+	var errs []error
+	// Validate cluster network config
+	if clusterConfig.Spec.NetworkType != types.NetworkType {
+		errs = append(errs, fmt.Errorf("expected network type is: %s, but got: %s", types.NetworkType, clusterConfig.Spec.NetworkType))
+	}
+	if len(clusterConfig.Spec.ServiceNetwork) == 0 {
+		errs = append(errs, fmt.Errorf("service network can not be empty"))
+	}
+	for _, network := range clusterConfig.Spec.ServiceNetwork {
+		if _, _, err := net.ParseCIDR(network); err != nil {
+			errs = append(errs, fmt.Errorf("invalid service network: %s", network))
+		}
+	}
+
+	// Validate antrea config
+	antreaAgentConfig := make(map[string]interface{})
+	err := yaml.Unmarshal([]byte(operConfig.Spec.AntreaAgentConfig), &antreaAgentConfig)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to parse %s option: %v", types.AntreaAgentConfigOption, err))
+	} else {
+		serviceCIDR, ok := antreaAgentConfig[types.ServiceCIDROption].(string)
+		if ok {
+			found := false
+			for _, serviceNet := range clusterConfig.Spec.ServiceNetwork {
+				if serviceNet == serviceCIDR {
+					found = true
+					break
+				}
+			}
+			if !found {
+				errs = append(errs, fmt.Errorf("invalid serviceCIDR option: %s, available values are: %s", serviceCIDR, clusterConfig.Spec.ServiceNetwork))
+			}
+		} else {
+			errs = append(errs, fmt.Errorf("serviceCIDR option can not be empty"))
+		}
+	}
+	if operConfig.Spec.AntreaImage == "" {
+		errs = append(errs, fmt.Errorf("antreaImage option can not be empty"))
+	}
+	if len(errs) > 0 {
+		return fmt.Errorf("invalidate configuration: %v", errs)
+	}
+	return nil
+}
+
+func NeedApplyChange(preConfig, curConfig *operatorv1.AntreaInstall) (bool, bool) {
+	if preConfig == nil {
+		return true, true
+	}
+
+	agentNeedChange := false
+	controllerNeedChange := false
+	if preConfig.Spec.AntreaAgentConfig != curConfig.Spec.AntreaAgentConfig {
+		agentNeedChange = true
+	}
+	if preConfig.Spec.AntreaCNIConfig != curConfig.Spec.AntreaCNIConfig {
+		agentNeedChange = true
+	}
+	if preConfig.Spec.AntreaControllerConfig != curConfig.Spec.AntreaControllerConfig {
+		controllerNeedChange = true
+	}
+	if preConfig.Spec.AntreaImage != curConfig.Spec.AntreaImage {
```

Comment:
K8s will restart pod if image is changed

Reply:
OK. I skip the `deleteExistingPods` if image is not get changed in update.